pkg/metrics/metric/metric.go

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metric
+
+import (
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// WithMetadata is the interface implemented by any metric defined in this package. These typically embed existing
+// prometheus metric types and add additional metadata to the definitions which can be used to generate proper metric
+// documentation from de definitions. In addition, these metrics have the concept of being enabled or disabled which
+// is used in place of conditional registration so all metric types can always be registered for the purposes of
+// documenting them.
+type WithMetadata interface {
+	IsEnabled() bool
+	SetEnabled(bool)
+	Opts() Opts
+	Labels() LabelDescriptions
+}
+
+// metric is a "base" structure which can be embedded to provide common functionality.
+type metric struct {
+	opts    Opts
+	enabled bool
+}
+
+func (b *metric) IsEnabled() bool {
+	return b.enabled
+}
+
+func (b *metric) SetEnabled(e bool) {
+	b.enabled = e
+}
+
+func (b *metric) Opts() Opts {
+	return Opts(b.opts)
+}
+
+func (b *metric) Labels() LabelDescriptions {
+	var labels LabelDescriptions
+	for constLabel, value := range b.opts.ConstLabels {
+		labels = append(labels, LabelDescription{
+			Name:        constLabel.Name,
+			Description: constLabel.Description,
+			KnownValues: []KnownValue{
+				{
+					Name: value,
+				},
+			},
+		})
+	}
+	return labels
+}
+
+// Opts are a modified and extended version of the prometheus.Opts
+// https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Opts
+type Opts struct {
+	// Namespace, Subsystem, and Name are components of the fully-qualified
+	// name of the Metric (created by joining these components with
+	// "_"). Only Name is mandatory, the others merely help structuring the
+	// name. Note that the fully-qualified name of the metric must be a
+	// valid Prometheus metric name.
+	Namespace string
+	Subsystem Subsystem
+	Name      string
+
+	// Help provides information about this metric.
+	//
+	// Metrics with the same fully-qualified name must have the same Help
+	// string.
+	//
+	// This string is included with the data in the prometheus endpoint and should be brief.
```

Comment:
Capitalize unless this is the name of a specific endpoint
```suggestion
	// This string is included with the data in the Prometheus endpoint and should be brief.
```